Chart Double Click

scallaway opened the pull request:
## Pre-ramble

This is in prep for a PR to go back into the main Recharts repository. I have a feeling that we're going to need to wait until after they complete the `beta` upgrade that they're in the process of doing at the moment, however.

Once that is complete, we can bump this repo to the latest master and then look to get a PR ready to be merged in.

## PR content

Adds the ability to pass an `onDoubleClick` prop to the charts in the same way that single clicks are implemented.

Also updates the `package-lock.json` file with the automatic changes after running an `npm install`.

scallaway-uipath committed:
Add Double Click to charts

Adds the ability to pass through an `onDoubleClick` prop to charts so that
callbacks can be written to handle with this use case separate from the
`onClick` prop.

kerron reviewed (approved):
This LGTM! 👍

richsilv commented on package-lock.json:
```diff
@@ -5664,7 +5664,8 @@
           "bundled": true,
-          "dev": true
+          "dev": true,
+          "optional": true
```
I'm not sure this is a good idea - they may be based on an older version of NPM and only want to upgrade in their own time (and in a separate PR). It feels like we're conflating concerns to an extent.

scallaway replied:
Okay that's a fair point. Hopefully this'll be addressed as part of their beta update process.

richsilv reviewed (changes requested):
The code looks good. It's hugely boring, but they have both tests and demos in the repo, and we should be adding examples of the new handler to both (probably only one to the latter) for it to stand a chance of getting merged.

scallaway commented:
@richsilv 

Have now updated the PR with some tests as well as examples in the demos (only very simple ones).

The actual test landscape wasn't great itself so I wasn't 100% sure how to deal with that, but I have added some regular `onClick` prop tests as well as the new `onDoubleClick` prop tests whether I thought they were necessary.

scallaway commented:
@richsilv 

Have now updated the PR with some tests as well as examples in the demos (only very simple ones).

The actual test landscape wasn't great itself so I wasn't 100% sure how to deal with that, but I have added some regular `onClick` prop tests as well as the new `onDoubleClick` prop tests whether I thought they were necessary.

richsilv commented on demo/component/BarChart.js:
```diff
@@ -239,6 +239,10 @@ export default class Demo extends Component {
   };
 
+  handleDoubleClick = () => {
+    console.log('Double click invoked!');
```
Could you also log the provided arguments here, just so the user can see them being passed through to the handler?

scallaway replied:
Yep, not sure why I didn't do that in the first place. Definitely worth showing that.

richsilv commented on demo/component/PieChart.js:
```diff
@@ -130,6 +130,10 @@ export default class Demo extends Component {
   };
 
+  handleDoubleClick = () => {
+    console.log('Double click invoked!');
```
As above.

richsilv commented on src/chart/Treemap.js:
```diff
@@ -398,6 +399,33 @@ class Treemap extends PureComponent {
+        currentRoot: root,
+        nestIndex,
+      });
+    }
```
I don't think we want to be doing any of this, because the `click` handler will already fire on the first click of a double-click, and that will do whatever this is doing (which I assume is zooming in or something). If anything, maybe we want to suppress a further zoom on the second click, but it's hard to know without an example. So:

1. Does this do something sensible? It's hard to know without a demo or test.
2. Is there even something sensible it could be doing in the context of a treemap, given the issues of a click always firing before a double-click?
3. Given that we're not using the treemap and it's meaningfully different from the other chart types, is this out of scope for what you're trying to achieve? If you add a simple demo and it doesn't act as you'd expect straight away, I'd argue so.

scallaway replied:
Yeah had a play around with the treemap and it does appear to be a much more complex case than the other charts. There would have to be some way of delaying the click just enough so that the double click could be detected without causing too much lag. Would be a load of work and I would agree that it's out of the scope of this change.

Going to back it out.

richsilv reviewed (changes requested):
This is really good, thanks. My main issue is that the treemap double-click isn't really tested or demonstrated, and might do strange things, so it may be best to leave it as is - it's up to you.

The other issue is the documentation (which obviously doesn't include onDoubleClick), but helpfully that's in a totally separate repo here, so I suggest you submit this PR and submit a separate one there if anybody ever responds to it and asks you to update the docs.

richsilv reviewed (approved):
LGTM, thanks!
